source/crud/bulk-write.md

```diff
@@ -0,0 +1,724 @@
+     */
+    hint: Optional<(String | Document)>;
+
+    /**
+     * When true, creates a new document if no document matches the query. Defaults to false.
+     *
+     * This options is sent only if the caller explicitly provides a value.
+     */
+    upsert: Optional<Boolean>;
+}
+
+class DeleteOneModel implements WriteModel {
+    /**
+     * The filter to apply.
+     */
+    filter: Document;
+
+    /**
+     * Specifies a collation.
+     *
+     * This option is sent only if the caller explicitly provides a value.
+     */
+    collation: Optional<Document>;
+
+    /**
+     * The index to use. Specify either the index name as a string or the index key pattern. If
+     * specified, then the query system will only consider plans using the hinted index.
+     *
+     * This option is only sent if the caller explicitly provides a value.
+     */
+    hint: Optional<(String | Document)>;
+}
+
+class DeleteManyModel implements WriteModel {
+    /**
+     * The filter to apply.
+     */
+    filter: Document;
+
+    /**
+     * Specifies a collation.
+     *
+     * This option is sent only if the caller explicitly provides a value.
+     */
+    collation: Optional<Document>;
+
+    /**
+     * The index to use. Specify either the index name as a string or the index key pattern. If
+     * specified, then the query system will only consider plans using the hinted index.
+     *
+     * This option is only sent if the caller explicitly provides a value.
+     */
+    hint: Optional<(String | Document)>;
+}
+```
+
+Each write model provided to `MongoClient.bulkWrite` in the `models` parameter MUST have a
+corresponding namespace that defines the collection on which the operation should be performed.
+Drivers SHOULD design this pairing in whichever way is most ergonomic for its language. For
+example, drivers may:
+
+- Include a required `namespace` field on each `WriteModel` variant.
+- Accept a list of `(Namespace, WriteModel)` tuples for `models`.
+- Define the following pair class:
+
+```typescript
+class NamespaceWriteModelPair {
+    /**
+     * The namespace on which to perform the write.
+     */
+    namespace: Namespace;
+
+    /**
+     * The write to perform.
+     */
+    model: WriteModel;
+}
+```
+
+Drivers MUST throw an exception if the list provided for `models` is empty.
+
+### Options
+
+```typescript
+class BulkWriteOptions {
+    /**
+     * Whether the operations in this bulk write should be executed in the order in which they were
+     * specified. If false, writes will continue to be executed if an individual write fails. If
+     * true, writes will stop executing if an individual write fails.
+     *
+     * Defaults to false.
+     */
+    ordered: Optional<Boolean>;
+
+    /**
+     * If true, allows the writes to opt out of document-level validation.
+     *
+     * Defaults to false.
+     *
+     * This option is only sent if the caller explicitly provides a value.
```

Comment:
FWIW, there doesn't appear to be any record of why the CRUD spec was modified to not sent `false` for `bypassDocumentValidation`. See: https://github.com/mongodb/specifications/pull/266#issuecomment-431491212

My last follow-up comment in [DRIVERS-450](https://jira.mongodb.org/browse/DRIVERS-450) (ticket created _after_ the spec change was merged) speculated that `bypassDocumentValidation: false` was unique in expressing a double-negative.

I could go either way here, but if we decide to keep the language as-is here, I'd suggest we also open a DRIVERS ticket to revise the CRUD spec and remove the ignorance of `false`. That can be a low priority change, as it's unlikely to affect an application and there's presently no test coverage anyway (see: [DRIVERS-1316](https://jira.mongodb.org/browse/DRIVERS-1316)).

Reply:
> I think we should leave this as-is.

Agreed. I am convinced to send what is set, and rely on the server to interpret the value. Suggest documenting in the Q&A section since it is a notable change with existing CRUD spec:

```md
### Why is `bypassDocumentValidation: false` sent?

`BulkWriteOptions.bypassDocumentValidation` is only sent if set (to either true or false).
`InsertOneOptions.bypassDocumentValidation` is only sent if set to true (not sent if false).

The rationale for only sending `InsertOneOptions.bypassDocumentValidation` when
true is not clear. DRIVERS-450 notes possible rationale.

Checking the values that users explicitly provide for options and omitting them
if they match the server default seems like unnecessary additional work for
drivers. It could also lead to issues in the unlikely event that the server
changes the default value for an option.
```

> speculated that `bypassDocumentValidation: false` was unique in expressing a double-negative.

I can see the double-negative being confusing. Though in the event `bypassDocumentValidation=false` is set and the user's intent is to opt-out of document validation, the result is doing an extra document validation (not harmful IMO).

Reply:
Added a Q&A entry and filed DRIVERS-2865